### PR TITLE
use `properToJs` and `identToJs` instead of `runProperName` to properly escape characters

### LINKED ISSUE #5: Replace single quotes in type names with a valid character for TS
As single quote `'` is a valid character in PS for type names, and it is invalid for TS. I think replacing the single quote with some other valid character is needed?
Examples of such types are in https://pursuit.purescript.org/packages/purescript-profunctor-lenses/3.2.0/docs/Data.Lens.Types

Not sure though which character is the best to replace it

Edit:
I have 2 proposed solutions:
1. Deal with it like Purescript itself converting any `'` to `$prime`

2. Replace it with a valid unicode. I found 2 unicode characters that would represent the single quote in TS in variable/type names while being considered the same as single quote in PS (so would throw a duplicate name error at compile time if used along with the single quote):
    - [Modified Letter Prime](https://www.compart.com/en/unicode/U+02B9): `ʹ`
    - [Modified Letter Apostrophe](https://www.compart.com/en/unicode/U+02BC): `ʼ`

    Try it in TS: https://www.typescriptlang.org/play/#code/C4TwDgpgBAgonUBQLxQN4EMBcUDOwBOAlgHYDmAvgNxQD01UYhAthALABQ7okUAQoD1ASVFEw58xElEo06aMAHtceOWAAWrNkA

### PR DESCRIPTION
**Pull request by sharno**
Fixes #5

### CODE REVIEW

File: src/Language/PureScript/TsdGen/Types.hs [hunk 1]
```diff
@@ -267,7 +267,7 @@ showFunctionParameters types = T.intercalate ", " $ zipWith (\n ty -> "_" <> T.p
 
 objectPropertyToString :: PSString -> Text
 objectPropertyToString ps = case decodeString ps of
-                              Just t | not (isValidJsIdentifier t) -> t
+                              Just t | not (isValidJsIdentifier t) -> anyNameToJs t
```
Comment (minoki) [inline src/Language/PureScript/TsdGen/Types.hs]: The use of `anyNameToJs` is inappropriate here.

PureScript compiles a function like

```purescript
someFn :: { foo' :: Int, bar :: String } -> Int
someFn r = r.foo'
```

into

```js
var someFn = function (r) {
    return r["foo'"];
};
```

, not

```js
var someFn = function (r) {
    return r.foo$prime;
};
```
.

The problem with `objectPropertyToString` is that the guard condition is reversed: the correct code is `Just t | isValidJsIdentifier t`.

I fixed this problem in df6645d382b5515183be1fda1b06b0cd75612131.

File: src/Language/PureScript/TsdGen/Module.hs [hunk 2]
```diff
@@ -190,13 +189,13 @@ processLoadedModule env ef importAll = execWriterT $ do
...
-                             in TSNamed Nothing (runProperName edTypeName <> "$$" <> runProperName ctorPName) (map TSTyVar fv)
+                             in TSNamed Nothing (name <> "$$" <> properToJs ctorPName) (map TSTyVar fv)
 
                          -- the data constructor is not exportd (i.e. abstract):
                          -- the marker fields are non-optional, so that they cannot be implicitly casted from other types.
                            | otherwise
-                           = TSRecord [ mkField "$$pursType" (TSStringLit $ mkString $ runModuleName currentModuleName <> "." <> runProperName edTypeName)
-                                      , mkField "$$pursTag" (TSStringLit $ mkString $ runProperName ctorPName)
+                           = TSRecord [ mkField "$$pursType" (TSStringLit $ mkString $ runModuleName currentModuleName <> "." <> name)
+                                      , mkField "$$pursTag" (TSStringLit $ mkString $ properToJs ctorPName)
```
Comment (minoki) [inline src/Language/PureScript/TsdGen/Module.hs]: The type of marker fields are string literals, so no escaping is needed. Of course, there would be no harm in changing it, but I'd rather keep it as is.

File: src/Language/PureScript/TsdGen/Module.hs [hunk 3]
```diff
@@ -248,13 +247,13 @@ processLoadedModule env ef importAll = execWriterT $ do
...
-                      dataCtorSubtypeName = runProperName edDataCtorTypeCtor <> "$$" <> runProperName edDataCtorName
+                      dataCtorSubtypeName = properToJs edDataCtorTypeCtor <> "$$" <> name
                       dataCtorSubtype = TSNamed Nothing dataCtorSubtypeName (map TSTyVar fieldTypeVars)
                   fieldTypesTS <- mapM (pursTypeToTSTypeX fieldTypeVars) fieldTypes
                   let mkMarkerField | length constructors == 1 = mkOptionalField -- allow structural subtyping if there are only one constructor
                                     | otherwise = mkField -- nominal typing
-                      makerFields = [ mkMarkerField "$$pursType" (TSStringLit (mkString $ runModuleName currentModuleName <> "." <> runProperName edDataCtorTypeCtor))
-                                    , mkMarkerField "$$pursTag" (TSStringLit (mkString $ runProperName edDataCtorName))
+                      makerFields = [ mkMarkerField "$$pursType" (TSStringLit (mkString $ runModuleName currentModuleName <> "." <> properToJs edDataCtorTypeCtor))
+                                    , mkMarkerField "$$pursTag" (TSStringLit (mkString $ name))
```
Comment (minoki) [inline src/Language/PureScript/TsdGen/Module.hs]: The type of marker fields are string literals, so no escaping is needed. Of course, there would be no harm in changing it, but I'd rather keep it as is.

File: src/Language/PureScript/TsdGen/Module.hs [hunk 4]
```diff
@@ -291,12 +290,12 @@ processLoadedModule env ef importAll = execWriterT $ do
...
           emitValueDeclaration (Just "instance") name instanceTy
       | otherwise = emitComment ("invalid instance declaration '" <> name <> "'")
-      where name = runIdent edInstanceName :: Text
+      where name = identToJs edInstanceName :: Text
             qDictTypeName = fmap coerceProperName edInstanceClassName :: Qualified (ProperName 'TypeName)
 
     processDecl EDKind{..} = do
       -- Do nothing for kind declarations: just put a comment.
-      let name = runProperName edKindName
+      let name = properToJs edKindName
```
Comment (minoki) [inline src/Language/PureScript/TsdGen/Module.hs]: Escaping here is not needed, because the name is used only in a comment.

**Review by minoki — COMMENTED** [review]
Some of the changes were unnecessary, but other changes look good to me.

File: src/Language/PureScript/TsdGen/Types.hs (hunk 1 above)
Comment (sharno) [reply src/Language/PureScript/TsdGen/Types.hs]: That sounds good, reverted it back
Thanks for the quick fix

**Comment by minoki** [comment]
Merged. Thanks!

**Pull request merged**